cranelift/codegen/src/machinst/lower.rs

```diff
@@ -1338,6 +1402,7 @@ impl<'func, I: VCodeInst> Lower<'func, I> {
                             + 1
                             == self.cur_scan_entry_color.unwrap().get()
                     {
+                        debug_assert!(!is_value_use_root(self.f, src_inst));
```

Comment:
This is asserting that (in effect) we don't have any multiple-result instructions that are side-effecting, I think. That may be true today but I wouldn't want to implicitly bake it in -- could we add `!is_value_use_root(...)` as an additional condition to this if-statement instead?

Reply:
I mostly wanted to make sure that the preexisting `self.num_outputs(src_inst) == 1` on this case didn't become accidentally load-bearing so I put this here to ensure that if `is_value_use_root` changed that it would alert that this needs to be changed as well. That being said I like your suggestion as well, so I've done a bit of refactoring to lift `is_value_use_root(...)` checking to the top of the conditional to handle both pure/impure cases at the same time.

Reply:
Nice!